File: csharp/src/Vanni.Idmix/Number.cs
using System.Text;

namespace Vanni.Idmix;

internal static class Number
{
    internal const int MaxStringLen = 63;

    internal sealed class DataObject
    {
        public bool IsString { get; init; }
        public int OType { get; init; }
        public long Val { get; init; }
        public byte[] Str { get; init; } = [];
    }

    internal static DataObject[] NormalizeObjects(object[] values)
    {
        var outArr = new DataObject[values.Length];
        for (var i = 0; i < values.Length; i++)
        {
            try
            {
                outArr[i] = ObjectFromAny(values[i]);
            }
            catch (Exception ex)
            {
                throw new ArgumentException($"value[{i}]: {ex.Message}", ex);
            }
        }
        return outArr;
    }

    internal static object[] MaterializeObjects(DataObject[] objects)
    {
        var outArr = new object[objects.Length];
        for (var i = 0; i < objects.Length; i++)
        {
            if (objects[i].IsString)
            {
                outArr[i] = Encoding.UTF8.GetString(objects[i].Str);
                continue;
            }
            outArr[i] = MaterializeValue(objects[i]);
        }
        return outArr;
    }

    internal static DataObject ObjectFromAny(object v)
    {
        switch (v)
        {
            case string s:
                if (s.Length == 0)
                    throw new ArgumentException($"empty string is not allowed (max {MaxStringLen} bytes)");
                var strBytes = Encoding.UTF8.GetBytes(s);
                if (strBytes.Length > MaxStringLen)
                    throw new ArgumentException($"string length {strBytes.Length} exceeds max {MaxStringLen}");
                return new DataObject { IsString = true, Str = strBytes };
            case byte[] bs:
                if (bs.Length == 0)
                    throw new ArgumentException($"empty byte slice is not allowed (max {MaxStringLen} bytes)");
                if (bs.Length > MaxStringLen)
                    throw new ArgumentException($"byte slice length {bs.Length} exceeds max {MaxStringLen}");
                var copy = new byte[bs.Length];
                Buffer.BlockCopy(bs, 0, copy, 0, bs.Length);
                return new DataObject { IsString = true, Str = copy };
            case TypedValue tv:
                return new DataObject { OType = tv.OType, Val = tv.Val };
            case byte x: return new DataObject { OType = TypedValue.OTypeUint8, Val = x };
            case ushort x: return new DataObject { OType = TypedValue.OTypeUint16, Val = x };
            case uint x: return new DataObject { OType = TypedValue.OTypeUint32, Val = x };
            case ulong x: return new DataObject { OType = TypedValue.OTypeUint64, Val = unchecked((long)x) };
            case sbyte x: return new DataObject { OType = TypedValue.OTypeInt8, Val = x };
            case short x: return new DataObject { OType = TypedValue.OTypeInt16, Val = x };
            case int x: return new DataObject { OType = TypedValue.OTypeInt32, Val = x };
            case long x: return new DataObject { OType = TypedValue.OTypeInt64, Val = x };
            default:
                throw new ArgumentException(
                    $"unsupported type {v.GetType().Name} (integer or string up to {MaxStringLen} bytes)");
        }
    }

    private static object MaterializeValue(DataObject obj) => obj.OType switch
    {
        TypedValue.OTypeUint8 => (byte)obj.Val,
        TypedValue.OTypeUint16 => (ushort)obj.Val,
        TypedValue.OTypeUint32 => (uint)obj.Val,
        TypedValue.OTypeUint64 => unchecked((ulong)obj.Val),
        TypedValue.OTypeInt8 => (sbyte)obj.Val,
        TypedValue.OTypeInt16 => (short)obj.Val,
        TypedValue.OTypeInt32 => (int)obj.Val,
        TypedValue.OTypeInt64 => obj.Val,
        _ => throw new ArgumentException($"invalid otype {obj.OType}"),
    };
}

File: java/src/main/java/io/github/vannifan/idmix/Number.java
package io.github.vannifan.idmix;

import java.nio.charset.StandardCharsets;

/** Type normalization for Idx encode/decode. */
final class Number {
    static final int MAX_STRING_LEN = 63;

    private Number() {}

    static final class DataObject {
        final boolean isString;
        final int otype;
        final long val;
        final byte[] str;

        private DataObject(boolean isString, int otype, long val, byte[] str) {
            this.isString = isString;
            this.otype = otype;
            this.val = val;
            this.str = str;
        }

        static DataObject integer(int otype, long val) {
            return new DataObject(false, otype, val, null);
        }

        static DataObject string(byte[] str) {
            return new DataObject(true, 0, 0, str);
        }
    }

    static DataObject[] normalizeObjects(Object[] values) {
        DataObject[] out = new DataObject[values.length];
        for (int i = 0; i < values.length; i++) {
            try {
                out[i] = objectFromAny(values[i]);
            } catch (IllegalArgumentException ex) {
                throw new IllegalArgumentException("value[" + i + "]: " + ex.getMessage(), ex);
            }
        }
        return out;
    }

    static Object[] materializeObjects(DataObject[] objects) {
        Object[] out = new Object[objects.length];
        for (int i = 0; i < objects.length; i++) {
            if (objects[i].isString) {
                out[i] = new String(objects[i].str, StandardCharsets.UTF_8);
            } else {
                out[i] = TypedValue.fromOtype(objects[i].otype, objects[i].val);
            }
        }
        return out;
    }

    static DataObject objectFromAny(Object v) {
        if (v instanceof String s) {
            if (s.isEmpty()) {
                throw new IllegalArgumentException("empty string is not allowed (max " + MAX_STRING_LEN + " bytes)");
            }
            byte[] bytes = s.getBytes(StandardCharsets.UTF_8);
            if (bytes.length > MAX_STRING_LEN) {
                throw new IllegalArgumentException("string length " + bytes.length + " exceeds max " + MAX_STRING_LEN);
            }
            return DataObject.string(bytes);
        }
        if (v instanceof byte[] bs) {
            if (bs.length == 0) {
                throw new IllegalArgumentException("empty byte slice is not allowed (max " + MAX_STRING_LEN + " bytes)");
            }
            if (bs.length > MAX_STRING_LEN) {
                throw new IllegalArgumentException("byte slice length " + bs.length + " exceeds max " + MAX_STRING_LEN);
            }
            byte[] copy = new byte[bs.length];
            System.arraycopy(bs, 0, copy, 0, bs.length);
            return DataObject.string(copy);
        }
        if (v instanceof TypedValue tv) {
            return DataObject.integer(tv.otype, tv.val);
        }
        if (v instanceof Byte x) return DataObject.integer(TypedValue.OTYPE_INT8, x);
        if (v instanceof Short x) return DataObject.integer(TypedValue.OTYPE_INT16, x);
        if (v instanceof Integer x) return DataObject.integer(TypedValue.OTYPE_INT32, x);
        if (v instanceof Long x) return DataObject.integer(TypedValue.OTYPE_INT64, x);
        throw new IllegalArgumentException(
                "unsupported type " + v.getClass().getSimpleName() + " (integer or string up to " + MAX_STRING_LEN + " bytes)");
    }

    private static Object materializeValue(DataObject obj) {
        return TypedValue.fromOtype(obj.otype, obj.val);
    }
}
